hw/opentitan/ot_usbdev.c

```diff
@@ -553,6 +821,116 @@ ot_usbdev_set_raw_link_state(OtUsbdevState *s, OtUsbdevLinkState state)
         FIELD_DP32(s->regs[R_USBSTAT], USBSTAT, LINK_STATE, (uint32_t)state);
 }
 
+/*
+ * Write the content of a USB buffer.
+ *
+ * @buf_id ID of the buffer
+ * @buf Buffer holding the data
+ * @size How much data to copy
+ */
+static void ot_usbdev_write_buffer(OtUsbdevState *s, uint8_t buf_id,
+                                   const uint8_t *data, size_t size)
+{
+    g_assert(size <= OT_USBDEV_MAX_PACKET_SIZE);
+    g_assert(buf_id < OT_USBDEV_BUFFER_COUNT);
+
+    /*
+     * See BFM (write_pkt_bytes).
+     *
+     * The buffer is stored as words, which means that if the packet size
+     * is not a multiple of 4, the RTL will zero out the bytes at the end
+     * of the last word.
+     *
+     * This function does not fully emulate the complex behaviour w.r.t
+     * the CRC being written to the buffer or data being remembered in
+     * the flops (see BFM).
+     */
+    uint32_t *ptr =
+        &s->buffer[buf_id * OT_USBDEV_MAX_PACKET_SIZE / sizeof(uint32_t)];
+    while (size >= 4u) {
+        *ptr++ = ldl_le_p(data);
+        data += sizeof(uint32_t);
+        size -= 4u;
+    }
+    /* Handle last incomplete word */
+    uint32_t val32 = 0u;
+    if (size >= 1u) {
+        val32 |= *data++;
+    }
+    if (size >= 2u) {
+        val32 |= *data++ << 8u;
+    }
+    if (size >= 3u) {
+        val32 |= *data++ << 16u;
+    }
+    if (size > 0u) {
+        *ptr++ = val32;
+    }
+}
+
+/*
+ * Read the content of a USB buffer.
+ *
+ * @buf_id ID of the buffer
+ * @buf Buffer where to copy the data
+ * @size How much data to copy
+ */
+static void ot_usbdev_read_buffer(OtUsbdevState *s, uint8_t buf_id,
+                                  uint8_t *buf, size_t size)
+{
+    g_assert(size <= OT_USBDEV_MAX_PACKET_SIZE);
+    g_assert(buf_id < OT_USBDEV_BUFFER_COUNT);
+
+    /*
+     * The buffer is stored as words, we need to emulate that. Could be
+     * optimized.
+     */
+    uint32_t *ptr =
+        &s->buffer[buf_id * OT_USBDEV_MAX_PACKET_SIZE / sizeof(uint32_t)];
+    while (size >= 4u) {
+        stl_le_p(buf, *ptr++);
+        buf += sizeof(uint32_t);
+        size -= 4u;
+    }
+    /* Handle last incomplete word */
+    if (size >= 1u) {
+        uint32_t val32 = *ptr;
+        *buf++ = val32 & 0xff;
```

Comment:
Nit: this should really be

```suggestion
        *buf++ = (uint8_t)val32;
```
(repeated twice below)